src/librbd/migration/NativeFormat.cc

```diff
@@ -0,0 +1,146 @@
+
+namespace librbd {
+namespace migration {
+
+template <typename I>
+NativeFormat<I>::NativeFormat(
+    I* image_ctx, const MigrationInfo& migration_info)
+  : m_image_ctx(image_ctx), m_migration_info(migration_info) {
+}
+
+template <typename I>
+void NativeFormat<I>::open(Context* on_finish) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 10) << dendl;
+
+  // TODO add support for external clusters
+  librados::IoCtx io_ctx;
+  int r = util::create_ioctx(m_image_ctx->md_ctx, "source image",
+                             m_migration_info.pool_id,
+                             m_migration_info.pool_namespace, &io_ctx);
+  if (r < 0) {
+    on_finish->complete(r);
+    return;
+  }
+
+  m_image_ctx->md_ctx.dup(io_ctx);
+  m_image_ctx->data_ctx.dup(io_ctx);
+
+  uint64_t flags = 0;
+  if (m_migration_info.image_id.empty()) {
+    m_image_ctx->name = m_migration_info.image_name;
+    flags |= OPEN_FLAG_OLD_FORMAT;
+  } else {
+    m_image_ctx->id = m_migration_info.image_id;
+  }
+
+  // set rados flags for reading the parent image
+  if (m_image_ctx->child->config.template get_val<bool>("rbd_balance_parent_reads")) {
+    m_image_ctx->set_read_flag(librados::OPERATION_BALANCE_READS);
+  } else if (m_image_ctx->child->config.template get_val<bool>("rbd_localize_parent_reads")) {
+    m_image_ctx->set_read_flag(librados::OPERATION_LOCALIZE_READS);
+  }
+
+  // open the source RBD image
+  auto ctx = util::create_async_context_callback(*m_image_ctx, on_finish);
+  m_image_ctx->state->open(flags, ctx);
+}
+
+template <typename I>
+void NativeFormat<I>::close(Context* on_finish) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 10) << dendl;
+
+  // the native librbd::image::CloseRequest handles all cleanup
+  on_finish->complete(0);
+}
+
+template <typename I>
+void NativeFormat<I>::get_snapshots(SnapInfos* snap_infos, Context* on_finish) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 10) << dendl;
+
+  m_image_ctx->image_lock.lock_shared();
+  *snap_infos = m_image_ctx->snap_info;
+  m_image_ctx->image_lock.unlock_shared();
+
+  on_finish->complete(0);
+}
+
+template <typename I>
+void NativeFormat<I>::get_image_size(uint64_t snap_id, uint64_t* size,
+                                     Context* on_finish) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 10) << dendl;
+
+  m_image_ctx->image_lock.lock_shared();
+  *size = m_image_ctx->get_image_size(snap_id);
+  m_image_ctx->image_lock.unlock_shared();
+
+
+  on_finish->complete(0);
+}
+
+template <typename I>
+void NativeFormat<I>::read(uint64_t snap_id, io::Extents&& image_extents,
+                           int op_flags, int read_flags, bufferlist* data,
+                           const ZTracer::Trace &parent_trace,
+                           Context* on_finish) {
+  auto cct = m_image_ctx->cct;
+  ldout(cct, 20) << "snap_id=" << snap_id << ", "
+                 << "image_extents=" << image_extents << dendl;
+  io::ReadResult read_result{data};
```

Comment:
@dillaman Do I understand right that we don't support sparse read here?
And the same for `RawFormat`?

Reply:
But probably not a big deal, if it is not going to be used by background migration (deep copy).

Reply:
Good point, I'll switch the API to just use a `ReadResult` to hide the underlying type -- and to allow `NativeFormat` to directly pass it and continue to utilize sparse bufferlists.